### PR TITLE
Stop liquids from flowing over ignore.

### PR DESCRIPTION
**Pull request by red-001**

### CODE REVIEW

**Comment by PilzAdam** [comment]
Keeping the nodes that are on top of unloaded blocks constantly in the queue is really bad.

**Comment by red-001** [comment]
PilzAdam then how do you update them when the block is loaded?

**Comment by PilzAdam** [comment]
@red-001 dunno, I guess thats the reason why it wasn't done before ;-)

Maybe add liquids on top of loaded blocks to the update queue? But then it would be quite inconsistent, since other liquids are not updated when loading a block...

**Comment by red-001** [comment]
PilzAdam I checked the memory and CPU usage with and without my modification and I couldn't see any difference. You could always add an option to disable this.

**Comment by PilzAdam** [comment]
@red-001 well, the CPU usage doesn't increase because the loopcount is limited by liquid_loop_max. The problem is that other liquid nodes will take longer until they are updated.

Also what was your scenario? Imagine a skyblock game where there are a lot of water and lava streams flowing down on unloaded blocks.

**Comment by red-001** [comment]
I tested it in skyblock with 9 water and 9 lava sources.
The update speed seemed to decrees only slightly in with my modification and the cpu usage seemed  to be the a slightly higher without it.

**Comment by kwolekr** [comment]
I don't think ignored flows need to be readded to the queue.  Any neighboring node that is set would cause that node to get re-added to the transforming liquid queue, which is coincidentally great for us because that node won't be able to flow until there is a change from content_ignore to something else, anyway.

So, when content_ignore is detected, just skip to the next item.

**Comment by red-001** [comment]
kwolekrI I tried that. It didn't work. Nodes aren't add to queue on block load.

**Comment by kwolekr** [comment]
Hrmm.  That's true.  Maybe you should try a different approach.

Here's what I had in mind:  When content_ignore is encountered in a neighbor, add that node position (or maybe block position, or both) to a queue.  We'll call it the transform-liquid-in-pending-block-load queue.  Next cycle, check if any blocks in that queue have been loaded.  If so, add the node position(s) associated with that block to the main transforming liquid queue and remove that block position from the pending load queue.

I realize it is much more complex than what you have, but continuously saturating the liquid transform queue is a pretty bad since it would prevent some nodes from ever getting transformed and it'll make half the liquids on the map "stuck" under some intensive situations, such as caves in Mapgen V5 with lava.

**Comment by red-001** [comment]
Wouldn't it be better to just update blocks on load?

**Comment by HybridDog** [comment]
>Wouldn't it be better to just update blocks on load?

It would make liquids generate map on skyblocks server, wouldn't it?

**Comment by PilzAdam** [comment]
@red-001 that may add too much load when people are walking around. Maybe an ABM would be better, although that would require a lot of tuning.

**Comment by paramat** [comment]
Anyone want to tackle this? It will be good for floatlands and using water columns for travelling between them.

File: src/map.cpp [hunk 1]
```diff
@@ -1731,7 +1732,11 @@ void Map::transformLiquids(std::map<v3s16, MapBlock*> & modified_blocks)
 						if (nb.t == NEIGHBOR_LOWER) {
 							flowing_down = true;
 						}
-					} else {
+					}
+					else if (nb.n.getContent() == CONTENT_IGNORE){
```
Comment (nerzhul) [inline src/map.cpp]: space

**Comment by Fixer-007** [comment]
IMO. This issue should have high priority as it is often found on actual MP servers and causes different kinds of water problems that is pain to resolve.

**Comment by gregorycu** [comment]
I think red-001's idea about updating on load is good. I understand the potential for a performance problem, however, I've made the lighting code much faster so we have some wiggle room. We should attempt this solution and observe the performance code.

I agree with kwolekr that saturating the transform queue is horrible. With regards to the transform-liquid-in-pending-block-load queue, that would have to be persisted.

**Comment by kwolekr** [comment]
Hmm you're right, if the pending queue isn't saved to disk then we could have the situation where liquid flowed into ignore, stopped, added that block to the queue, then the server restarts and leaves off with an empty queue so that block is never updated for liquids.  Darn.  I had an otherwise solid plan :(

I agree that updating the whole block (or at least the nodes along the edges...) would be the easiest thing to do but that's at the very least 96 nodes added to the queue per load.  Too much for my comfort.

**Comment by kwolekr** [comment]
Then again, the liquid transform queue itself has no persistence, so what is it exactly that makes it okay for that situation?

**Comment by gregorycu** [comment]
Maybe it suffers the same fate? Could be a bug that nobody has reported?

**Comment by kwolekr** [comment]
It's possible.  In any case, I'm adding a scan along the edges for liquids in activateBlock().  It'll add any it finds to the transform queue.

This should be a really good balance between performance and implementation simplicity.

**Comment by gregorycu** [comment]
I think so too.

**Comment by kwolekr** [comment]
No good, I tried that strategy and it keeps the number of items in the liquid transforming queue at around ~12000 when map is loading near ground level.  I'm not really comfortable with such an impact for a small edge case like this.

**Comment by gregorycu** [comment]
I don't understand the maths, to be honest. How can it keep things in the
queue?

**Comment by gregorycu** [comment]
Sorry. It adds new things to the queue constantly to keep it at that level.
I get it now.

**Comment by paramat** [comment]
Thanks for the atttempt. I'm thinking perhaps an ABM that occasionally updates nodes (including falling nodes to make the hangng sand created by mapgen collapse, something i was considering anyway). However the way liquid spreads out on ignore should be fixed first because if the whole area of spread water starts flowing down it will look even more ugly and multiply the problem.

**Comment by Fixer-007** [comment]
Is it possible to activate when player is near in some vicinity (like check for hanging water in 100 blocks or whatever).

**Comment by red-001** [comment]
@paramat spreading over ignore is easy to fix. This PR already does that, just remove [this line](https://github.com/minetest/minetest/pull/2977/files#diff-ab005ad9757130721ffb66a09833090aR1781) and you have the water just stop when it reaches ignore. An ABM could be used to restart it again but that would be feel a bit like a hack that barely works. Water is not done very well in general (e.g place a source block high in the air and then quit and you get water stuck in the air and not flowing)  and needs a lot of work to make it work semi-realistically.

**Comment by paramat** [comment]
red-001 thanks i'll try that except i'll only continue when the node below is ignore.

**Comment by paramat** [comment]
See #3581 for an experiment.

**Comment by paramat** [comment]
kwolekr > No good, I tried that strategy and it keeps the number of items in the liquid transforming queue at around ~12000 when map is loading near ground level.

How about adding to the liquid queue only the flowing liquids in the lowest node layer of a mapblock, on block load? This will cover most cases of stuck liquids. not adding sources avoids huge amounts of sea being added.
We need an engine solution, FIxer and i have been experimenting with an ABM but it doesn't work too well and is heavy.

**Comment by Fixer-007** [comment]
I have a world copy with lots of broken flowing liquids hanging in the air, when I start it freshly near that liquid... closest ones somehow reactivate itself, but if I start far from them - they hang. I'm not sure if this reproducible, or maybe I did something here. Also, if my memory is not wrong 99.9999% of liquid hang is flowing kind (without any sources at all).

If some one wants to experience some real bugged liquids, here is the world (go south from spawn to reach increasingly big amounts of them): [Download: Just Test.zip](https://github.com/minetest/minetest/files/101456/Just.Test.zip)

**Comment by red-001** [comment]
Maybe do what @paramat say and add an option like "fix_liquids" that will update **all** the liquids on block load allowing badly bugged maps to be fixed without mods?

**Comment by paramat** [comment]
Actually there is already an API to add a node to the liquid update queue, but it didn't seems to work for me, maybe it wan't acting on the lowest node in a column:
```minetest.transforming_liquid_add(pos)```

> when I start it freshly near that liquid... closest ones somehow reactivate itself, but if I start far from them - they hang

Maybe because ABMs have an active range of 32 nodes?

**Comment by Fixer-007** [comment]
```minetest.transforming_liquid_add(pos)```
does not working for me for upper and bottom water flowing nodes, no activation :/

**Comment by paramat** [comment]
Closing as my PR was merged. Discussion of water reflow methods is useful.

**Comment by MillersMan** [comment]
I've created a request (#4061) for a water reflow implementation that will solve this.

**Pull request closed without merging**